OTA-1339: `UpdateStatus`: API to expose information about update progress & health

petr-muller opened the pull request:
Alternative, smaller-footprint version of the API proposed in https://github.com/openshift/api/pull/2012, discussed in https://github.com/openshift/enhancements/pull/1701 , addressing feedback on the early [Update Health API Controller Proposal](https://docs.google.com/document/d/1SuV8mvEQbUMpcV1VTazuPlXgbMeUm2134SIV5so4a5M/edit) and [Update Health API Draft](https://docs.google.com/document/d/1aEIWkfhhaVSe-XlSXvmokymOe3X_969pRCJhfhPDwFQ/edit) docs.

This alternative proposal completely removes the "structure" part of `UpdateStatus` singleton proposed in https://github.com/openshift/api/pull/2012, and extracts the "data" part of it ("insight" leaves of the `UpdateStatus` tree) to a top-level, individual, smaller-footprint CRDs. This approach completely removes some of the concerns voiced in https://github.com/openshift/api/pull/2012. I have some concerns about the approach (mostly about losing functionality), but I am able to iterate and maybe the structure part can be addressed separately.

Except for the high-level layer, the [example walkthrough](https://github.com/openshift/enhancements/blob/7687d60346060818d7f4725ef0d4529ca197b26b/enhancements/update/update-status-api/update-status-api-example.md) (why the data structures are shaped like this and what functionality they are supposed to support) is applicable here as well.

/cc @JoelSpeed @sdodson @wking

petr-muller commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+// +kubebuilder:resource:path=clusterversionprogressinsights,scope=Cluster
+// +openshift:api-approved.openshift.io=https://github.com/openshift/api/pull/2012
+// +openshift:file-pattern=cvoRunLevel=0000_00,operatorName=cluster-version-operator,operatorOrdering=02
+// +openshift:enable:FeatureGate=UpgradeStatus
```
`UpgradeStatus` is a TechPreview featuregate. At this point I think we'd need to have a new, `DevPreview` featuregate if this is accepted.

petr-muller replied:
xref: https://github.com/openshift/api/pull/2246

wking commented on update/v1alpha1/types_clusteroperator_progress_insight.go:
```diff
@@ -0,0 +1,133 @@
+	// Progressing is used with Updating=True when the ClusterOperator is updating
+	ClusterOperatorUpdatingReasonProgressing ClusterOperatorUpdatingReason = "Progressing"
+	// CannotDetermine is used with Updating=Unknown
+	ClusterOperatorUpdatingCannotDetermine ClusterOperatorUpdatingReason = "CannotDetermine"
```
nit: there doesn't seem to be any enum-style validation on `conditions[].reason`, so maybe we can leave these off and go free-form in the resource-creating controller?  We'd need `Updated` or `Pending` for the `Updating=False` case to allow clients to distinguish there, but I'm not clear on why clients would need to distinguish by well-known reason for `Updating=True` or `Updating=Unknown`, especially with us only declaring one `reason` string for each of those `status` values.  But also, obviously not a big deal if we wanted central definitions of some default reasons, in case that helps make resource-creators more consistent in the "there's no more-granular information we want to share" cases.

petr-muller replied:
I'd more inclined to add validations eventually rather than drop to free-form, because I think we should either be full "reasons are well-known and can matter" or full  "reasons are free-form", but not half-way. I'll add the CEL rules.

petr-muller replied:
I have started adding validations for conditions (to ensure condition presence and that they use the correct reasons etc) but then I realized tightening the API this much at this point makes it harder to experiment in the controller (for example, I'm not sure if I can get away with a single `Healthy` condition on CV & MCP insights, or whether I would need to break it down. For now the API targets DevPreview and does not need to be this tight, so for now I only added TODOs everywhere.

wking commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+// ControlPlaneUpdateVersions contains the original and target versions of the upgrade
+type ControlPlaneUpdateVersions struct {
+	// previous is the version of the control plane before the update. When the cluster is being installed
+	// for the first time, the version will have a placeholder value '<none>' and carry 'Installation' metadata
```
nit: if the cluster history is:

1. Completed vA.
2. Partial vB.
3. Partial vC.

it's not clear to me from these docs whether the value would be vA or vB.  [The client-side code suggests vB][1], in which case maybe wording like:

> previous is the version the control plane was attempting to reconcile before the update.

or some such that makes it clear that it's "the previous target, regardless of completion" and not "the most recent completed target".

[1]: https://github.com/openshift/oc/blob/fe18ae25a123e896fd218210073141f7f5811465/pkg/cli/admin/upgrade/status/controlplane.go#L416-L418

petr-muller replied:
Improved the godoc, hopefully

wking commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+
+// ScopeType is one of ControlPlane or WorkerPool
+// +kubebuilder:validation:Enum=ControlPlane;WorkerPool
+type ScopeType string
```
nit: `ScopeType` doesn't seemed to be consumed by `ClusterVersionProgressInsight`.  Maybe move it to `types_health_insight.go` to live alongside the consuming structure?

petr-muller replied:
It's also used in MCP and Node insights. Introducing a "common.go" felt like overhead but maybe it's worth it (especially if we keep using `ResourceRef` everywhere and not just in Health)

petr-muller replied:
Moved to `types_common.go`

wking commented on update/v1alpha1/types_node_progress_insight.go:
```diff
@@ -0,0 +1,161 @@
+	// +required
+	Scope ScopeType `json:"scopeType"`
+
+	// version is the version of the node, when known
```
nit: this isn't making previous/target distinctions.  Did we want to recycle something like `ControlPlaneUpdateVersions`?  But obviously not with that kind of name specificity, we'd want `UpdateVersions` or `NodeUpdateVersions` or some such, depending on whether we thought we could share with the control plane consumer or not.

If all we need is a single version string, can we expand the Godocs to be clear about whether it's previous or target version?  I'm guessing it's target version?

petr-muller replied:
It is the current version - previous before the node is updated, and target afterwards.

petr-muller replied:
https://github.com/openshift/enhancements/blob/7687d60346060818d7f4725ef0d4529ca197b26b/enhancements/update/update-status-api/update-status-api-example.md#node-status-insight tries to help

petr-muller replied:
Clarified in godoc

wking reviewed (approved):
I left some nit comments inline, but nothing stood out to me as a structural issue (not surprising given the amount of time you've invested in polishing this API proposal so far).  So approving, because Godocs and that level thing can all get sorted out in post-merge follow-up, if folks can find a way to land dev-preview APIs so we have time to build the tooling that writes and reads these while Godocs and things that the tooling won't care about get sorted.

petr-muller commented on update/v1alpha1/types_node_progress_insight.go:
```diff
@@ -0,0 +1,161 @@
+	// +than type safety for producers.
+	// +required
+	// +kubebuilder:validation:XValidation:rule="!has(self.group) && self.resource == 'nodes'",message="resource must be a nodes.core.k8s.io resource"
+	Resource ResourceRef `json:"resource"`
```
Yeah, technically `Name` is enough because progress insight types are coupled with exactly one kind, at least for now (Nodes were a thing I suspected may benefit from having a reference to e.g. a `Machine` and the new MCO nodeupdatestatus).

Consistency for clients is part of the attempted value, and each insight type using consistent reference structs whenever they refer to other resources felt useful, so that if clients want to do some "read insight then read actual resource" logic, they do not need to implement the translation themselves for each insight type but just read `.resource` and feed it into some shared utility.

JoelSpeed replied:
Joining from another thread, I also noted that this seems redundant.

If you think there's value in the resource (specifying group and resource type), I don't really see the value in the name since it's already in the resource reference.

You've mentioned the reason for having a reference, but not why name as well?

petr-muller replied:
I am not _entirely_ convinced, but it's not the hill I need to die on so in the spirit of simplification I can sacrifice the intended convenience, and will gut the `.resource` fields :+1:

JoelSpeed commented on update/v1alpha1/types_machineconfigpool_progress_insight.go:
```diff
@@ -0,0 +1,173 @@
+	// summaries is a list of counts of nodes matching certain criteria (e.g. updated, degraded, etc.). Maximum 16 items can be listed.
+	// +listType=map
+	// +listMapKey=type
+	// +patchStrategy=merge
+	// +patchMergeKey=type
```
Patch/merge strategy doesn't actually work on CRDs

JoelSpeed replied:
Yeah they should be removed from everywhere, the linter is slightly out of date, I'll get that updated and let you know when that's done

JoelSpeed commented on update/v1alpha1/types_machineconfigpool_progress_insight.go:
```diff
@@ -0,0 +1,173 @@
+	// +kubebuilder:validation:Maximum=100
+	Completion int32 `json:"completion"`
+
+	// summaries is a list of counts of nodes matching certain criteria (e.g. updated, degraded, etc.). Maximum 16 items can be listed.
```
Since the type is an enum, the maximum should be tied to the number of different possible enum values

JoelSpeed commented on update/v1alpha1/types_machineconfigpool_progress_insight.go:
```diff
@@ -0,0 +1,173 @@
+	// +required
+	Type NodeSummaryType `json:"type"`
+
+	// count is the number of nodes matching the criteria, between 0 and 1024
```
I suspect we support more than 1024 nodes, perhaps make the limit 5000? Probably want to check with perf and scale, but I believe our limit is well below 5000

petr-muller replied:
https://docs.redhat.com/en/documentation/openshift_container_platform/4.18/html/scalability_and_performance/planning-your-environment-according-to-object-maximums#cluster-maximums-major-releases_object-limits

says 2k

JoelSpeed replied:
Shall we set the limit to 2k then?

petr-muller replied:
Why not

JoelSpeed commented on update/v1alpha1/types_node_progress_insight.go:
```diff
@@ -0,0 +1,161 @@
+	// +required
+	Scope ScopeType `json:"scopeType"`
+
+	// version is the version of the node, when known
```
What version is this? And, does it make sense to tie this to the Kube/OpenShift semver (looks like semver-ish match?) when Nodes are actually tied to a rendered MC version, and that can change over time even when the cluster OCP version stays the same?

petr-muller replied:
From the user PoV it makes sense - if you start updating from 4.17 to 4.18 then seeing nodes bump from 4.17 to 4.18 is what you expect. "tied to a rendered MC version" is an implementation detail that experts know.

JoelSpeed replied:
Ack, context in the godoc would be helpful here

JoelSpeed reviewed (commented):
I'm not doing a fine tooth comb review at the moment because I know we are keen to get something in and iterate, I'm working out a good way to make sure the API review is completed before we GA, and track that as part of the work, but for now, some initial thoughts

petr-muller commented on update/v1alpha1/types_clusteroperator_progress_insight.go:
```diff
@@ -0,0 +1,133 @@
+// component update in standalone clusters), during the update
+// +kubebuilder:validation:XValidation:rule="self.name == self.resource.name",message=".name must match .resource.name"
+type ClusterOperatorProgressInsightStatus struct {
+	// conditions provide details about the operator. It contains at most 10 items. Known conditions are:
+	// - Updating: whether the operator is updating; When Updating=False, the reason field can be Pending or Updated
+	// - Healthy: whether the operator is considered healthy; When Healthy=False, the reason field can be Unavailable or Degraded, and Unavailable is "stronger" than Degraded
+	// +listType=map
+	// +listMapKey=type
+	// +patchStrategy=merge
+	// +patchMergeKey=type
+	// +optional
+	// +kubebuilder:validation:MaxItems=10
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```
It is hard for cluster operators to implement these conditions because they require awareness of state beyond the ClusterOperator `spec->status` reconciliation (specifically the `Updating` one). `Updating=False Reason=Pending` is not a state you can detect if you only look at the CO, you need the knowledge that the CVO started updating the cluster, and therefore that the CO _should_ be eventually bumped.

But yeah, this one is probably the thinnest of the status insights.

petr-muller commented on update/v1alpha1/types_clusteroperator_progress_insight.go:
```diff
@@ -0,0 +1,133 @@
+	// +kubebuilder:validation:Pattern=`^[a-z0-9-]+$`
+	Name string `json:"name"`
+
+	// resource is the ClusterOperator resource that represents the operator
+	//
+	// +Note: By OpenShift API conventions, in isolation this should be a specialized reference that refers just to
+	// +resource name (because the rest is implied by status insight type). However, because we use resource references in
+	// +many places and this API is intended to be consumed by clients, not produced, consistency seems to be more valuable
+	// +than type safety for producers.
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.group == 'config.openshift.io' && self.resource == 'clusteroperators'",message="resource must be a clusteroperators.config.openshift.io resource"
+	Resource ResourceRef `json:"resource"`
```
https://github.com/openshift/api/pull/2233#discussion_r1998860947 responds to this.

petr-muller commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+	// +patchMergeKey=type
+	// +optional
+	// +kubebuilder:validation:MaxItems=10
+	Conditions []metav1.Condition `json:"conditions,omitempty" patchStrategy:"merge" patchMergeKey:"type"`
```
The whole purpose of Status API is for consumers to not "just look at the cluster version". ClusterVersion has multiple conditions that are essentially phases (`ReleaseAccepted`, `Upgradeable`, `Progressing`). Right now the implementation mimics `Progressing` (cvo actually loaded new payload, validated it, and started reconciling) but in the future I wanted to cover the states before `Progressing=True` in this insight. So again the CV insight is a result of a logic of interpreting a potentially complicated CV state.

petr-muller commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+	// +many places and this API is intended to be consumed by clients, not produced, consistency seems to be more valuable
+	// +than type safety for producers.
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.group == 'config.openshift.io' && self.resource == 'clusterversions'",message="resource must be a clusterversions.config.openshift.io resource"
```
I'm not sure whether the API needs to encode the knowledge of  this convention. Whatever produces the insights is responsible for putting correct data there.

JoelSpeed replied:
What would an insight achieve if it pointed to ClusterVersion called foo? There would be no `foo` ClusterVersion of that name, so, does it make sense to even allow that?

petr-muller commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+	// +kubebuilder:validation:XValidation:rule="self.group == 'config.openshift.io' && self.resource == 'clusterversions'",message="resource must be a clusterversions.config.openshift.io resource"
+	Resource ResourceRef `json:"resource"`
+
+	// assessment is the assessment of the control plane update process. Valid values are: Unknown, Progressing, Completed, Degraded
+	// +required
+	Assessment ControlPlaneAssessment `json:"assessment"`
```
Multiple conditions need to be interpreted, and that is logic. The purpose of this API is for client to not need the interpretation logic. This field is not supposed to be a machine-readable Enum to e.g. switch on, but human-oriented brief summary of the state. It is not expected to be switched-on, but rather fed into the UI. I guess the documentation should stress that out more?

JoelSpeed replied:
Additional docs would help here yes

petr-muller replied:
I added a paragraph of text that tries to describe the intent (also to the other assessment-y field on MCP)

petr-muller commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+	// +required
+	Versions ControlPlaneUpdateVersions `json:"versions"`
+
+	// completion is a percentage of the update completion (0-100)
+	// +required
+	// +kubebuilder:validation:Minimum=0
+	// +kubebuilder:validation:Maximum=100
+	Completion int32 `json:"completion"`
```
That means the last update (or "update that is actually installation") is complete and this should be 100

JoelSpeed replied:
Worth adding docs explaining this as it may not be obvious

petr-muller replied:
Added godocs for this (also on other fields)

petr-muller commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+	// previous is the version of the control plane before the update. When the cluster is being installed
+	// for the first time, the version will have a placeholder value '<none>' and carry 'Installation' metadata
+	// +required
+	// +kubebuilder:validation:XValidation:rule="self.version == '<none>' ? (has(self.metadata) && self.metadata.exists(m, m.key == 'Installation')) : !(has(self.metadata) && self.metadata.exists(m, m.key == 'Installation'))",message="previous version must be '<none>' iff marked with Installation metadata"
```
I guess it is a matter of explicit (previous is empty and it explicitly tells you why) vs implied (which needs to be interpreted, encoding the knowledge).

I think I originally mostly avoided optional `Previous` because of OpenShift conventions saying "don't use pointers" so I can imagine dropping this.

JoelSpeed replied:
Not using pointers doesn't mean not using optional fields, the two are disconnected

I would have said that an empty `previous` or absent `previous` is fairly clear that there's been no previous version. As long as the rest of the status looks like it's been successfully updated I would think that's a fairly intuitive interpretation

JoelSpeed replied:
There's a lot of nuance around when something should or shouldn't be a pointer.

I think you're starting from the wrong side of the decision process.

First up, decide if you want your field optional or required, once you've made that decision, we can talk about pointers and omitempty.

Choose optional vs required based on the actual expected usage of the field.

This is a very required heavy status, which is unusual, APIs usually have mostly optional status fields

If you decide the field is definitely required, it can stay as it is.

If you decide the field is optional, you should add omitempty.

If you add omitempty, that will only work on structs when they are pointers.

You only _need_ the omitempty to work correctly when there are required fields within the struct.

If the field is optional, and there are no required fields within it, then non-pointer, non-omitempty is also a fine way to represent this, and in theory improves discoverability (though that's not important for status generally)

petr-muller commented on update/v1alpha1/types_health_insight.go:
```diff
@@ -0,0 +1,181 @@
+	// +required
+	Type ScopeType `json:"type"`
+
+	// resources is a list of resources involved in the insight, of any group/kind. Maximum 16 resources can be listed.
```
Good question. I think pushing some scaling limits is better suited for a future development because right now I do not see an immediate case where this would be needed. To me these limits would be mostly a contract for the insight producer to comply to ("what if insight summary / description needs to be longer than allowed" is a similar kind of question), I guess the guidance for now would be to put the most relevant ones there.

JoelSpeed replied:
I think clarifying the expectation within the godoc would be important before committing to this limit.

I would personally opt for a higher limit, say 128, unless you think a `get -o yaml` will become unreadable with that many resources?

JoelSpeed commented on update/v1alpha1/types_health_insight.go:
```diff
@@ -0,0 +1,181 @@
+	// +required
+	Level InsightImpactLevel `json:"level"`
+
+	// type is the type of the impact. Valid values are None, Unknown, API Availability, Cluster Capacity,
```
> then the consumer would need a PascalCase -> Pascal Case

Why? What is the consumer here? Are we suggesting users of this won't understand how to read these values without spaces?

petr-muller replied:
They will understand, but why make them worse for humans when the values target humans? Human-oriented content is not uncommon in APIs, we allow messages and other free form text.

petr-muller commented on update/v1alpha1/types_clusterversion_progress_insight.go:
```diff
@@ -0,0 +1,263 @@
+	// +patchStrategy=merge
+	// +patchMergeKey=key
+	// +optional
+	// +kubebuilder:validation:MaxItems=5
```
Actually 2 because `Installation` is always target, and `Partial` is always previous

petr-muller commented:
OK tests pass with this revision: https://prow.ci.openshift.org/view/gs/test-platform-results/pr-logs/pull/openshift_api/2233/pull-ci-openshift-api-master-integration/1902789916102758400

Now I will push the version that adds a CEL validation for progress insights to enforce `.metadata.name == .status.name`, which makes all tests fail for me locally, and I cannot figure out why.

petr-muller commented:
/test minor-images

```
 INFO[2025-03-21T14:44:21Z] build didn't start running within 1h0m0s (phase: Pending):
* Container docker-build is not ready with reason PodInitializing 
```
Not me!

petr-muller commented:
@JoelSpeed I think I addressed all feedback now, PTAL. I have filed https://github.com/openshift/api/pull/2246 to create a new FG for this effort, and once it is approved, I will update it here as well. And once the PR is close to approval, I will squash the commits to something reasonable.

petr-muller commented:
/test all
